### ISSUE TITLE
run_on_batch takes target data as part of other arguments

### ISSUE THREAD

**Issue by mcgibbon** [ai2cm/ace#271]
See https://github.com/ai2cm/full-model/pull/507/files#r1476829738

**Comment by gideonite** [comment]
Here's what I think we should do:

1. Renaming `run_on_batch` to train on batch
2. Implementing a `forward` method so that this becomes a valid torch Module.

**Comment by mcgibbon** [comment]
> Here's what I think we should do:
> 
> 1. Renaming `run_on_batch` to train on batch
> 2. Implementing a `forward` method so that this becomes a valid torch Module.

1. `run_on_batch` is still used in our inference code to provide the same data alignment code to both training and inference when reporting metrics to wandb, this won't be changing even if we support a mode to call our stepper without data targets.

2. What is the motivation to make the stepper a torch module? Torch modules support a very wide API and are quite complex. The boundary of stepper containing a torch module while not being one itself has been quite useful. Everything related to neural net logic is isolated within the Module, while everything related to data pipelining , metrics, and optimization is outside of a Module. It's not clear what it means for a torch module to receive an optimization object as an argument, and it's very unusual for it to compute a loss and call the optimizer in its forward function. Is there an advantage we gain by making the stepper a Module?

I would propose we separate out the forward prediction component of `run_on_batch` into a `predict` method, which will be used by the existing `run_on_batch` (functionality unchanged). This would give us a way to call the stepper without target data, while run_on_batch would continue to align the target data with predictions, normalize it, and package it as we expect for giving to the Aggregators. run_on_batch would also continue to handle calling the optimization.

**Comment by mcgibbon** [comment]
I think there could be a follow-on opportunity to do something like what you've suggested, by pulling the `predict` method into a forward function on a wrapper module to apply over the built module, which extends it to run for multiple steps. The Stepper itself would still be a non-module, but you might be able to use the StepperModule subset for certain use cases. I'm not exactly clear on whether this is possible/reasonable or what advantage it would give, but it should be clearer after refactoring to have more "module-like" method such as `predict`.

**Comment by oliverwm1** [comment]
> I would propose we separate out the forward prediction component of run_on_batch into a predict method, which will be used by the existing run_on_batch (functionality unchanged).

This is also what I had in mind. For the "ACE" case, we could also go further and have "predict" and "predict_single_step" where the latter just steps forward one time step. I think this could be useful for other users who perhaps want to implement the inference in a different way.

**Comment by oliverwm1** [comment]
Related: ai2cm/ace#258

**Comment by gideonite** [comment]
Sounds good, I will move forward with implementing a predict method for the downscaling code.

**Comment by oliverwm1** [comment]
I made some progress in refactoring `run_on_batch` to use functions which do not require the target data here: https://github.com/ai2cm/full-model/pull/603
